homeassistant/components/tuya/__init__.py

```diff
@@ -191,31 +242,98 @@ async def async_unload_entry(hass, entry):
         )
     )
     if unload_ok:
+        for listener in hass.data[DOMAIN][entry.entry_id]["listener"]:
+            listener()
+        hass.data[DOMAIN].pop(entry.entry_id)
         hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
         hass.data[DOMAIN][TUYA_TRACKER]()
         hass.data[DOMAIN][TUYA_TRACKER] = None
         hass.data[DOMAIN][TUYA_DATA] = None
+        hass.data[DOMAIN].pop(TUYA_DEVICES_CONF)
         hass.services.async_remove(DOMAIN, SERVICE_FORCE_UPDATE)
         hass.services.async_remove(DOMAIN, SERVICE_PULL_DEVICES)
         hass.data.pop(DOMAIN)
 
     return unload_ok
 
 
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Update when config_entry options update."""
+    hass.data[DOMAIN][TUYA_DEVICES_CONF] = entry.options.copy()
+    _update_discovery_interval(
+        hass, entry.options.get(CONF_DISCOVERY_INTERVAL, DEFAULT_DISCOVERY_INTERVAL)
+    )
+    _update_query_interval(
+        hass, entry.options.get(CONF_QUERY_INTERVAL, DEFAULT_QUERY_INTERVAL)
+    )
+    async_dispatcher_send(hass, SIGNAL_CONFIG_ENTITY)
+
+
+async def cleanup_device_registry(hass: HomeAssistant, device_id):
+    """Remove device registry entry if there are no remaining entities."""
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    entity_registry = await hass.helpers.entity_registry.async_get_registry()
+    if device_id and not hass.helpers.entity_registry.async_entries_for_device(
+        entity_registry, device_id
+    ):
+        device_registry.async_remove_device(device_id)
+
+
 class TuyaDevice(Entity):
     """Tuya base device."""
 
+    _device_count = 0
+
     def __init__(self, tuya, platform):
         """Init Tuya devices."""
         self._tuya = tuya
         self._tuya_platform = platform
 
+    def _inc_device_count(self):
+        """Increment static variable device count."""
+        dev_type = self._tuya.device_type()
+        if dev_type in ["scene", "switch"]:
+            return
```

Comment:
Since this code is the same as the code in `_dec_device_count`, how about breaking this out into something like `self._device_uses_discovery`?? I'm not sure that name makes sense as I'm not 100% clear on the intent here.

Reply:
Called `self._device_can_use_query`